src/logging.rs

```diff
@@ -0,0 +1,64 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use log::{LevelFilter, Metadata, Record};
+use std::ffi::CString;
+use std::os::raw::c_char;
+use std::os::raw::c_int;
+
+pub type LogCallback = unsafe extern "C" fn(level: c_int, msg: *const c_char);
+
+struct LogProxy {
+    callback: LogCallback,
+}
+
+impl log::Log for LogProxy {
+    fn enabled(&self, _metadata: &Metadata) -> bool {
+        true
+    }
+
+    fn log(&self, record: &Record) {
+        if self.enabled(record.metadata()) {
+            let callback: LogCallback = self.callback;
+            let msg = record.args().to_string();
+            let msg_c = CString::new(msg).unwrap();
+            unsafe {
+                callback(record.level() as c_int, msg_c.as_ptr());
+            }
+            // We do not use std::mem::forget(s), so Rust will reclaim the memory
+            // once msg_c gets cleared. The callback should thus make a copy.
+        }
+    }
+
+    fn flush(&self) {}
+}
+
+// Store the logger instance as a mutable nullable static object
+static mut LOGGER: Option<LogProxy> = None;
+
+// Note: this function may only be called once in the lifetime of a program.
+#[no_mangle]
+pub unsafe extern "C" fn wgpu_set_log_callback(callback: LogCallback) {
+    // Instantiate logger, store as static object, and make it THE logger.
+    let logger: LogProxy = LogProxy { callback: callback };
```

Comment:
If you run the clippy tool on this file, it would suggest using:
```rust
let logger: LogProxy = LogProxy { callback };
```

Reply:
> If you run the clippy tool ...

```
λ rustup component add clippy
error: component 'clippy' for target 'x86_64-pc-windows-msvc' is unavailable for download
```
:(